module/builder/consensus/builder.go

```diff
@@ -232,12 +234,43 @@ func (b *Builder) BuildOn(parentID flow.Identifier, setter func(*flow.Header) er
 	// roadmap (https://github.com/dapperlabs/flow-go/issues/4872)
 
 	// create a mapping of block to seal for all seals in our pool
-	byBlock := make(map[flow.Identifier]*flow.Seal)
-	for _, seal := range b.sealPool.All() {
-		byBlock[seal.Seal.BlockID] = seal.Seal
+	// We consider two seals as inconsistent, if they have different start or end states
+	encounteredInconsistentSealsForSameBlock := false
+	byBlock := make(map[flow.Identifier]*flow.IncorporatedResultSeal)
+	for _, irSeal := range b.sealPool.All() {
+		if len(irSeal.IncorporatedResult.Result.Chunks) < 1 {
+			return nil, fmt.Errorf("ExecutionResult without chunks: %v", irSeal.IncorporatedResult.Result.ID())
+		}
+		if len(irSeal.Seal.FinalState) < 1 {
+			// respective Execution Result should have been rejected by matching engine
+			return nil, fmt.Errorf("seal with empty state commitment: %v", irSeal.ID())
+		}
+		if irSeal2, found := byBlock[irSeal.Seal.BlockID]; found {
+			sc1json, err := json.Marshal(irSeal)
+			if err != nil {
+				return nil, err
+			}
+			sc2json, err := json.Marshal(irSeal2)
+			if err != nil {
+				return nil, err
+			}
+
+			// check whether seals are inconsistent:
+			if !bytes.Equal(irSeal.Seal.FinalState, irSeal2.Seal.FinalState) ||
+				!bytes.Equal(irSeal.IncorporatedResult.Result.Chunks[0].StartState, irSeal2.IncorporatedResult.Result.Chunks[0].StartState) {
+				fmt.Printf("ERROR: inconsistent seals for the same block %v: %s and %s", irSeal.Seal.BlockID, string(sc1json), string(sc2json))
+				encounteredInconsistentSealsForSameBlock = true
+			} else {
+				fmt.Printf("WARNING: multiple seals with different IDs for the same block %v: %s and %s", irSeal.Seal.BlockID, string(sc1json), string(sc2json))
+			}
```

Comment:
I think this hotfix was more of a sanity check because the network was running several versions of the execution node at the time. Going forward I think it would be better to fail fast if we encounter either of these cases, since they represent a serious flaw in the execution node.

Reply:
Are you suggesting these to be fatal failure conditions, leading to node termination? 

* This would mean that the consensus node would just reboot with wiped seals mempool. Thereby it might potentially:
   - loose the conflicting seals
   - Upon re-requesting, there is a chance that only one of the conflicting seals is in the mempool first and we continue with sealing. In case the node again puts two conflicting seals in the mempool, it will just keep looping through the restart cycle ... until it eventually makes it past the point by luck.  
    - This would imply:
        - we have a fork in the chain of execution results
        - consensus nodes will eventually probably continue sealing -- non-deterministically selecting one of the two execution branches (i.e. the one which the sealing block proposer happened to know about)
* Essentially, a node failure is somewhat similar to just wiping the caches and trying to make progress on sealing again

In contrast, the current solution will keep the node running, with the inconsistent seals in the mempool
- during each block building attempt, the node will come across the conflicting seals and therefore _not_ include any seals: https://github.com/onflow/flow-go/blob/436c7b8c3be24885db8935ed3cc9435cb3e7adcc/module/builder/consensus/builder.go#L271-L274
- if all nodes have learned about the conflicting seals, sealing will halt -- which I think is what we desire (?)

What do you think Jordan? Does that make sense? Did I miss something?

Reply:
You're right, we shouldn't crash here. What you have makes sense 👌